drivers/platform/aducm3029/flash.c

```diff
@@ -0,0 +1,303 @@
+ *                          data length must be FLASH_PAGE_SIZE_WORDS in size.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t flash_write_page(struct flash_dev *dev, int32_t page_no, uint32_t *data)
+{
+	ADI_FEE_TRANSACTION transaction;
+	struct adicup_flash_dev *adicup_extra = dev->extra;
+	uint32_t fee_hw_error;
+	int32_t ret;
+
+	transaction.bUseDma = true;
+	transaction.nSize = FLASH_PAGE_SIZE_BYTES;
+	transaction.pWriteAddr = (uint32_t *)(page_no << FLASH_PAGE_ADDR_SHIFT);
+	transaction.pWriteData = data;
+
+	ret = adi_fee_Write(adicup_extra->instance, &transaction,
+			    &fee_hw_error);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * Read-modify-write helper function for flash_write(). Since the smallest write
+ * unit for the ADuCM3029 is the page this function helps access data smaller than
+ * a page or accross multiple pages.
+ */
+static int32_t flash_read_then_write(struct flash_dev *dev, uint32_t flash_addr,
+				     uint32_t *array, uint32_t array_size)
+{
+	ADI_FEE_TRANSACTION transaction;
+	struct adicup_flash_dev *adicup_extra = dev->extra;
+	uint32_t page_nr, fee_hw_error;
+	uint32_t temp_ptr[FLASH_PAGE_SIZE_WORDS] __attribute__ ((aligned (4)));
+	int32_t ret;
+
+	if ((flash_addr & 0x3) != 0)
+		return FAILURE;
+	if ((FLASH_ADDRESS_TO_PAGE(flash_addr) + (array_size * sizeof(uint32_t))) >
+	    0x800)
+		return FAILURE;
+
+	flash_read(dev, (flash_addr & (~FLASH_PAGE_START_ADDRESS_MASK)), temp_ptr,
+		   FLASH_PAGE_SIZE_WORDS);
+
+	memcpy(temp_ptr + FLASH_ADDRESS_TO_PAGE(flash_addr) / 4, array, array_size);
+
+	transaction.bUseDma = true;
+	transaction.nSize = FLASH_PAGE_SIZE_BYTES;
+	transaction.pWriteAddr = (uint32_t *)(flash_addr &
+					      (~FLASH_PAGE_START_ADDRESS_MASK));
+	transaction.pWriteData = temp_ptr;
+
+	/* Get the page number */
+	ret = adi_fee_GetPageNumber(adicup_extra->instance, flash_addr,
+				    &page_nr);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	/* First erase page */
+	ret = adi_fee_PageErase(adicup_extra->instance, page_nr, page_nr,
+				&fee_hw_error);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	/* Then write */
+	ret = adi_fee_Write(adicup_extra->instance, &transaction,
+			    &fee_hw_error);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * Write data in flash memory.
+ *
+ * @param [in] dev        - Pointer to the flash device handler.
+ * @param [in] flash_addr - Start address of the write.
+ * @param [in] array      - Pointer to the data to be written.
+ * @param [in] array_size - Size of the written data in 32-bit words.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t flash_write(struct flash_dev *dev, uint32_t flash_addr,
+		    uint32_t *array, uint32_t array_size)
+{
+	uint32_t data_index = 0, temp, initial_size = array_size;
+	int32_t ret;
+
+	do {
+		temp = (FLASH_PAGE_SIZE_BYTES - FLASH_ADDRESS_TO_PAGE(flash_addr)) / 4;
```

Comment:
It is strange to me that you substract from bytes number of pages. Something seems not to be ok here or am I not understanding correctly?

Reply:
Moved the division inside macro and I think I made it overall more consistent. What do you think?